src/cryptonote_core/service_node_list.h

```diff
@@ -378,27 +404,22 @@ namespace service_nodes
     cryptonote::BlockchainDB      *m_db;
 
     using block_height = uint64_t;
-    struct transient_state
+    struct state_t
     {
-      service_nodes_infos_t                                       service_nodes_infos;
-      std::vector<key_image_blacklist_entry>                      key_image_blacklist;
-      std::map<block_height, quorum_manager>                      quorum_states;
-      std::list<std::unique_ptr<rollback_event>>                  rollback_events;
-      block_height                                                height;
+      service_nodes_infos_t                  service_nodes_infos;
+      std::vector<key_image_blacklist_entry> key_image_blacklist;
+      block_height                           height;
 
       // Returns a filtered, pubkey-sorted vector of service nodes that are active (fully funded and
       // *not* decommissioned).
       std::vector<pubkey_and_sninfo> active_service_nodes_infos() const;
       // Similar to the above, but returns all nodes that are fully funded *and* decommissioned.
       std::vector<pubkey_and_sninfo> decommissioned_service_nodes_infos() const;
     };
-    transient_state m_transient_state;
 
-    template <typename T>
-    void load_transient(const rollback_event_variant &event)
-    {
-      m_transient_state.rollback_events.emplace_back(new T(boost::get<T>(event)));
-    }
+    std::map<block_height, quorum_manager> m_quorum_states;
+    std::deque<state_t>                    m_state_history;
+    state_t                               *m_state;
```

Comment:
Hmm the invalid state problem is still here if m_state_history is empty, this doesn't really solve it except hide it behind the accessor. Even adding a emplace_back if its empty at the accessor could be considered incorrect depending on when this function is called, it being empty could be considered valid (like at initialisation) or invalid (any other time where you expect there to be atleast one entry and if it is, we want a rescan to rebuild our now missing service node list state).

The other idea I toyed with is just declaring m_state by value and push the states into the history. The only downside is it makes serialization/reorg code need 2 paths for the deque and the current m_state when saving/clearing out history and I took the other as a trade off for this, but since you flagged it and seems to be a good middle ground. 

Actually not so bad having written it out, what do you think? Have updated the commit